source/material_model/drucker_prager.cc

```diff
@@ -29,56 +31,134 @@ namespace aspect
+                                              ?
+                                              reference_strain_rate * reference_strain_rate
+                                              :
+                                              std::fabs(second_invariant(strain_rate_deviator))));
+
+              const double sqrt3 = std::sqrt(3.0);
+
+              double strain_rate_effective = edot_ii_strict;
+
+              // plasticity
+              const double sin_phi = std::sin(angle_of_internal_friction);
+              const double cos_phi = std::cos(angle_of_internal_friction);
+              const double strain_rate_effective_inv = 1/(2*std::sqrt(strain_rate_effective));
+              const double strength_inv_part = 1/(sqrt3*(3.0+sin_phi));
+
+              const double strength = ( (dim==3)
+                                        ?
+                                        ( 6.0 * cohesion * cos_phi + 2.0 * pressure * sin_phi) * strength_inv_part
+                                        :
+                                        cohesion * cos_phi + pressure * sin_phi );
+
+              // Rescale the viscosity back onto the yield surface
+              const double eta_plastic = strength * strain_rate_effective_inv;
+
+
+              const double effective_viscosity = 1.0 / ( ( 1.0 / ( eta_plastic + minimum_viscosity ) ) + ( 1.0 / maximum_viscosity ) );
+
+              Assert(dealii::numbers::is_finite(effective_viscosity),ExcMessage ("Error: Viscosity is not finite."));
+              if (derivatives != NULL)
+                {
+                  if (std::sqrt(strain_rate_effective) > std::numeric_limits<double>::min())
+                    {
+                      const double averaging_factor = effective_viscosity * effective_viscosity * std::pow(eta_plastic + minimum_viscosity,-2);
+                      effective_viscosity_strain_rate_derivatives = averaging_factor * 0.5 * (eta_plastic * (-1/std::sqrt(edot_ii_strict * edot_ii_strict))) * strain_rate_deviator;
+                      effective_viscosity_pressure_derivatives = averaging_factor * (dim == 2 ?
+                                                                                     sin_phi * strain_rate_effective_inv
+                                                                                     :
+                                                                                     (sin_phi*strength_inv_part));
+                    }
+                  else
+                    {
+                      effective_viscosity_strain_rate_derivatives = 0;
+                    }
+                }
+
+              if (std::sqrt(strain_rate_effective) <= std::numeric_limits<double>::min())
+                {
+                  out.viscosities[i] = maximum_viscosity;
+                }
+              else
+                {
+                  out.viscosities[i] = effective_viscosity;
```

Comment:
I suppose since the tests pass the viscosity calculation is unchanged? I did not go through all of the formulas, but did you intentionally change something, or should everything be as before? Except for the addition of derivatives of course.